### ISSUE TITLE
Engine became unresponsive after interacting with Swarm API

### ISSUE THREAD

**Issue by alexellis** [docker/for-mac#1195]
### Expected behavior

Continued use of Docker

### Actual behavior

> Error response from daemon: Bad response from Docker engine

Restarting fails to correct the problem and I have to completely wipe my library to recover (it seems)

### Information

  - Full output of the diagnostics from "Diagnose & Feedback" in the menu
  - A reproducible case if this is a bug, Dockerfiles FTW
  - Page URL if this is a docs issue or the name of a man page

F222BA08-22D8-4914-9C31-FA0CFF14CF0C

### Steps to reproduce the behavior

This is hard - I think it was due to code in while updating replicas of a service programatically.

https://github.com/alexellis/faas/blob/59ca597903654afa577b24362f09ca6800369ee7/gateway/server.go

cc @justincormack

**Comment by samoht** [comment]
Not sure exactly what is going on, but you seem to be using both IPv6 and IPv4. That should normally be fine, but maybe could you try to disable IPv6 to see if that makes thing more reliable (just a wild guess..).

**Comment by samoht** [comment]
Hum actually, the docker engine seemed to have panicked:

```
Jan 23 22:30:12 moby root: time="2017-01-23T22:30:12.837824792Z" level=debug msg="Service yi2bew0kka6v5k8p2vrlstjxx was scaled down from 1 to -4 instances"
Jan 23 22:30:12 moby root: panic: runtime error: slice bounds out of range
Jan 23 22:30:12 moby root: goroutine 240 [running]:
Jan 23 22:30:12 moby root: panic(0x176af20, 0xc4200120f0)
Jan 23 22:30:12 moby root: ^I/usr/local/go/src/runtime/panic.go:500 +0x1a1
Jan 23 22:30:12 moby root: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated.(*Orchestrator).reconcile(0xc421a872c0, 0x250f820, 0xc420ab94c0, 0xc421059ce0)
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated/services.go:158 +0xca9
Jan 23 22:30:12 moby root: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated.(*Orchestrator).tickServices(0xc421a872c0, 0x250f820, 0xc420ab94c0)
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated/services.go:73 +0xd0
Jan 23 22:30:12 moby root: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated.(*Orchestrator).tick(0xc421a872c0, 0x250f820, 0xc420ab94c0)
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated/replicated.go:107 +0x61
Jan 23 22:30:12 moby root: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated.(*Orchestrator).Run(0xc421a872c0, 0x250f820, 0xc420ab94c0, 0x0, 0x0)
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/orchestrator/replicated/replicated.go:85 +0x3d5
Jan 23 22:30:12 moby root: github.com/docker/docker/vendor/github.com/docker/swarmkit/manager.(*Manager).becomeLeader.func10(0x250f820, 0xc420ab94c0, 0xc421a872c0)
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/manager.go:842 +0x46
Jan 23 22:30:12 moby root: created by github.com/docker/docker/vendor/github.com/docker/swarmkit/manager.(*Manager).becomeLeader
Jan 23 22:30:12 moby root: ^I/go/src/github.com/docker/docker/vendor/github.com/docker/swarmkit/manager/manager.go:845 +0x1163
```

@aaronl will maybe know more if that's related to your issue :-)

**Comment by samoht** [comment]
I have escalated your issue on docker/swarmkit: https://github.com/docker/swarmkit/issues/1906

**Comment by aaronlehmann** [comment]
Hi @alexellis - I opened https://github.com/docker/swarmkit/pull/1907 to fix this. I think the issue happens because `Replicas` is an unsigned field, so your `if replicas <= 0 {` check is ineffective. The field ends up underflowing, and this later leads to a panic inside the replicated orchestrator. In the PR, I've added validation at the API level that the value of `Replicas` does not look like an integer underflow, and also fixed the replicated orchestrator so it should avoid panicing in this case.
